cpp/src/experimental/louvain.cuh

```diff
@@ -34,21 +447,1824 @@ class Louvain {
+      dst_cache_v.resize(current_graph_view_.get_number_of_local_adj_matrix_partition_cols());
+    CUDA_TRY(cudaStreamSynchronize(stream_));
+      copy_to_adj_matrix_col(
+        handle_, current_graph_view_, local_input_v.begin(), dst_cache_v.begin());
+    }
+
+#ifdef DEBUG
+    barrier("end of cache_vertex_properties");
+    sleep(rank_);
+    printf("rank = %d\n", (int)rank_);
+    print_v("src_cache_v", src_cache_v);
+    print_v("dst_cache_v", dst_cache_v);
+#endif
+  }
+
+  //
+  // TODO CHECK:  Should be correct, calls other functions.
+  //              Does populate src_cluster_cache_v, dst_cluster_cache_v_ multiple
+  //              times and will update cluster_v_ if a better clustering is
+  //              identified
+  //
+  virtual weight_t update_clustering(weight_t total_edge_weight, weight_t resolution)
+  {
+    timer_start("update_clustering");
+
+    rmm::device_vector<vertex_t> next_cluster_v(cluster_v_);
+
+#ifdef DEBUG
+    barrier("update_clustering, calling cache_vertex_properties");
+    sleep(rank_);
+    print_v("next_cluster_v", next_cluster_v);
+    print_v("src_cluster_cache_v_", src_cluster_cache_v_);
+    print_v("dst_cluster_cache_v_", dst_cluster_cache_v_);
+#endif
+
+    cache_vertex_properties(next_cluster_v, src_cluster_cache_v_, dst_cluster_cache_v_);
+
+#ifdef DEBUG
+    barrier("update_clustering, cache_vertex_properties");
+    sleep(rank_);
+    print_v("next_cluster_v", next_cluster_v);
+    print_v("src_cluster_cache_v_", src_cluster_cache_v_);
+    print_v("dst_cluster_cache_v_", dst_cluster_cache_v_);
+#endif
+
+    weight_t new_Q = modularity(total_edge_weight, resolution);
+    weight_t cur_Q = new_Q - 1;
+
+#ifdef DEBUG
+    std::cout << "update_clustering, new_Q = " << new_Q << std::endl;
+#endif
+
+    // To avoid the potential of having two vertices swap clusters
+    // we will only allow vertices to move up (true) or down (false)
+    // during each iteration of the loop
+    bool up_down = true;
+
+    while (new_Q > (cur_Q + 0.0001)) {
+      cur_Q = new_Q;
+
+      update_by_delta_modularity(total_edge_weight, resolution, next_cluster_v, up_down);
+
+      up_down = !up_down;
+
+      cache_vertex_properties(next_cluster_v, src_cluster_cache_v_, dst_cluster_cache_v_);
+
+      new_Q = modularity(total_edge_weight, resolution);
+
+#ifdef DEBUG
+      if (rank_ == 0) std::cout << "new_Q = " << new_Q << std::endl;
+#endif
+
+      if (new_Q > cur_Q) {
+        thrust::copy(rmm::exec_policy(stream_)->on(stream_),
+                     next_cluster_v.begin(),
+                     next_cluster_v.end(),
+                     cluster_v_.begin());
+      }
+    }
+
+    // cache the final clustering locally on each cpu
+    cache_vertex_properties(cluster_v_, src_cluster_cache_v_, dst_cluster_cache_v_);
+
+    timer_stop(stream_);
+    return cur_Q;
+  }
+
+  //
+  // TODO CHECK:  This is the big one to debug in MNMG
+  //
+  void update_by_delta_modularity(weight_t total_edge_weight,
```

Comment:
Perhaps consider splitting this function into several separate ones (it's ~600 lines long; even though it includes debug logic seems a bit much)

Reply:
Will address this in 0.17.